src/Root16.Sprout/Extensions/IEnumerableExtensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Root16.Sprout.DependencyInjection;
+
+namespace Root16.Sprout.Extensions;
+
+public static class IEnumerableExtensions
+{
+    public async static Task<List<T>> ToListAsync<T>(this IAsyncEnumerable<T> values)
+    {
+        var returnedValues = new List<T>();
+        await foreach (var value in values)
+        {
+            returnedValues.Add(value);
+        }
+        return returnedValues;
+    }
+
+    internal async static IAsyncEnumerable<T> StreamFinishedTasksAllRunning<T>(this IList<Task<T>> tasks)
+    {
+        while (tasks.Any())
+        {
+            var finishedTask = await Task.WhenAny(tasks);
+            tasks.Remove(finishedTask);
+            yield return await finishedTask;
+        }
+    }
+
+    internal async static IAsyncEnumerable<TOutput> StreamFinishedTasksWithSpecificAmount<TOutput>(this List<KeyValuePair<StepRegistration,Func<StepRegistration,Task<TOutput>>>> tasksToRun, int amount = default)
+    {
+        var runningTasks = new List<Task<TOutput>>();
+        var newTasks = new List<KeyValuePair<StepRegistration, Func<StepRegistration, Task<TOutput>>>>();
+        //Start a specific amount of tasks and then remove them from the list of tasks to run
+        var TasksToRun = tasksToRun.Take(amount).ToList();
```

Comment:
@jrpetersjr - so nothing is starting the tasks in this method, I believe they are all already running.  Can we just wait for them all in the Task.WhenAny below instead of manipulating this partial list?

Reply:
The tasks are started on line 34. This gets the specific amount that's passed in to start at first because we might only want to run 2 at a time or we can run 7 at a time. 

I may be misunderstanding the question though.

Reply:
@csobrien - I added some comments and I updated the code a little bit